aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,564 @@
+//!
+//! // Alice recieves ciphertext bytes from bob
+//! let ciphertext_bytes = ciphertext.as_ref();
+//!
+//! // Bob sends Alice the ciphertext computed from the encapsulation algorithm, Alice runs decapsulation to derive their
+//! // copy of the secret.
+//! let alice_secret = decapsulation_key.decapsulate(Ciphertext::from(ciphertext_bytes))?;
+//!
+//! // Alice and Bob have now arrived to the same secret
+//! assert_eq!(alice_secret.as_ref(), bob_secret.as_ref());
+//!
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+use crate::{
+    error::{KeyRejected, Unspecified},
+    ptr::LcPtr,
+    ptr::Pointer,
+};
+use aws_lc::{
+    EVP_PKEY_CTX_kem_set_params, EVP_PKEY_CTX_new, EVP_PKEY_CTX_new_id, EVP_PKEY_decapsulate,
+    EVP_PKEY_encapsulate, EVP_PKEY_get_raw_private_key, EVP_PKEY_get_raw_public_key,
+    EVP_PKEY_kem_new_raw_public_key, EVP_PKEY_kem_new_raw_secret_key, EVP_PKEY_keygen,
+    EVP_PKEY_keygen_init, EVP_PKEY, EVP_PKEY_KEM,
+};
+use std::{borrow::Cow, cmp::Ordering, fmt::Debug, ptr::null_mut};
+use zeroize::Zeroize;
+
+/// An identifier for a KEM algorithm.
+pub trait AlgorithmIdentifier:
+    Copy + Clone + Debug + PartialEq + crate::sealed::Sealed + 'static
+{
+    /// Returns the algorithm's associated AWS-LC nid.
+    fn nid(self) -> i32;
+}
+
+/// A KEM algorithm
+#[derive(PartialEq)]
+pub struct Algorithm<Id = AlgorithmId>
+where
+    Id: AlgorithmIdentifier,
+{
+    id: Id,
+    secret_key_size: usize,
+    public_key_size: usize,
+    ciphertext_size: usize,
+    shared_secret_size: usize,
+}
+
+impl<Id> Algorithm<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    #[allow(dead_code)]
+    pub(crate) const fn new(
+        id: Id,
+        secret_key_size: usize,
+        public_key_size: usize,
+        ciphertext_size: usize,
+        shared_secret_size: usize,
+    ) -> Self {
+        Self {
+            id,
+            secret_key_size,
+            public_key_size,
+            ciphertext_size,
+            shared_secret_size,
+        }
+    }
+
+    /// Returns the identifier for this algorithm.
+    #[must_use]
+    pub fn id(&self) -> Id {
+        self.id
+    }
+
+    #[inline]
+    pub(crate) fn secret_key_size(&self) -> usize {
+        self.secret_key_size
+    }
+
+    #[inline]
+    pub(crate) fn public_key_size(&self) -> usize {
+        self.public_key_size
+    }
+
+    #[inline]
+    pub(crate) fn ciphertext_size(&self) -> usize {
+        self.ciphertext_size
+    }
+
+    #[inline]
+    pub(crate) fn shared_secret_size(&self) -> usize {
+        self.shared_secret_size
+    }
+}
+
+impl<Id> Debug for Algorithm<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self.id, f)
+    }
+}
+
+/// A serializable private key usable with KEMs. This can be randomly generated with `KemPrivateKey::generate`
+/// or constructed from raw bytes.
+pub struct DecapsulationKey<Id = AlgorithmId>
```

Comment:
Would be good if this could be `Sync` and `Send` if that is possible.

```
error[E0277]: `*mut aws_lc_sys::linux_x86_64_crypto::evp_pkey_st` cannot be shared between threads safely
   --> rustls-postquantum/src/lib.rs:83:28
    |
83  | impl ActiveKeyExchange for Active {
    |                            ^^^^^^ `*mut aws_lc_sys::linux_x86_64_crypto::evp_pkey_st` cannot be shared between threads safely
    |
    = help: within `Active`, the trait `Sync` is not implemented for `*mut aws_lc_sys::linux_x86_64_crypto::evp_pkey_st`
note: required because it appears within the type `ManagedPointer<*mut evp_pkey_st>`
   --> /home/jbp/.cargo/git/checkouts/aws-lc-rs-815c4a1fbe3a9dc5/de40172/aws-lc-rs/src/ptr.rs:18:19
    |
18  | pub(crate) struct ManagedPointer<P: Pointer> {
    |                   ^^^^^^^^^^^^^^
note: required because it appears within the type `KemPrivateKey`
   --> /home/jbp/.cargo/git/checkouts/aws-lc-rs-815c4a1fbe3a9dc5/de40172/aws-lc-rs/src/kem.rs:163:12
    |
163 | pub struct KemPrivateKey {
    |            ^^^^^^^^^^^^^
note: required because it appears within the type `Active`
   --> rustls-postquantum/src/lib.rs:77:8
    |
77  | struct Active {
    |        ^^^^^^
note: required by a bound in `ActiveKeyExchange`
   --> /home/jbp/rustls/rustls/src/crypto/mod.rs:268:37
    |
268 | pub trait ActiveKeyExchange: Send + Sync {
    |                                     ^^^^ required by this bound in `ActiveKeyExchange`

error[E0277]: `*mut aws_lc_sys::linux_x86_64_crypto::evp_pkey_st` cannot be sent between threads safely
   --> rustls-postquantum/src/lib.rs:83:28
    |
83  | impl ActiveKeyExchange for Active {
    |                            ^^^^^^ `*mut aws_lc_sys::linux_x86_64_crypto::evp_pkey_st` cannot be sent between threads safely
```

(I have picked up a different version of this PR that I rebased on top of main, so the type names are the old ones.)

Reply:
Good catch this was a miss, these API types should be Send and Sync. Thanks!